lib/web_ui/dev/roll_fallback_fonts.dart

```diff
@@ -509,3 +494,354 @@ bool _checkForLicenseAttribution(Uint8List fontBytes) {
+/// kinds of digit.  Lets say we use decimal digits `0`..`9` for prefix digits
+/// and `A`..`Z` as terminating digits.
+///
+///     M = ('M' - 'A') = 12
+///     38M = (3 * 10 + 8) * 26 + 12 = 38 * 26 + 12 = 1000
+///
+/// Choosing a large terminating radix is especially effective when most of the
+/// encoded values are small, as is the case with delta-encoding.
+///
+/// There can be multiple terminating digit kinds to represent different sorts
+/// of values. For the range table, the size uses a different terminating digit,
+/// 'a'..'z'. This allows the very common size of 1 (accounting over a third of
+/// the range) sizes to be omitted. A range is encoded as either
+/// `<size><value>`, or `<value>` with an implicit size of 1.  Since the size 1
+/// can be implicit, it is always implicit, and the stored sizes are biased by
+/// -2.
+///
+/// | encoding | value | size |
+/// | :---     | ---:  | ---: |
+/// | A        | 0     | 1    |
+/// | B        | 1     | 1    |
+/// | 38M      | 1000  | 1    |
+/// | aA       | 0     | 2    |
+/// | bB       | 1     | 3    |
+/// | zZ       | 25    | 27   |
+/// | 1a1A     | 26    | 28   |
+/// | 38a38M   | 1000  | 1002 |
+///
+/// STMR-encoded strings are decoded efficiently by a simple loop that updates
+/// the current value and performs some additional operation for a terminating
+/// digit, e.g. recording the optional size, or creating a range.
+
+String _computeEncodedFontSets(List<_Font> fonts) {
+  final List<_Range> ranges = <_Range>[];
+  final List<_FontSet> allSets = <_FontSet>[];
+
+  {
+    // The fonts have their supported code points provided as list of inclusive
+    // [start, end] ranges. We want to intersect all of these ranges and find
+    // the fonts that overlap each intersected range.
+    //
+    // It is easier to work with the boundaries of the ranges rather than the
+    // ranges themselves. The bounaries of the intersected ranges is the union
+    // of the boundaries of the individual font ranges.  We scan the boundaries
+    // in increasing order, keeping track of the current set of fonts that are
+    // in the current intersected range.  Each time the boundary value changes,
+    // the current set of fonts is canonicalized and recorded.
+    //
+    // There has to be a wiki article for this algorithm but I didn't find one.
+    final List<_Boundary> boundaries = <_Boundary>[];
+    for (final _Font font in fonts) {
+      for (final int start in font.starts) {
+        boundaries.add(_Boundary(start, true, font));
+      }
+      for (final int end in font.ends) {
+        boundaries.add(_Boundary(end + 1, false, font));
+      }
+    }
+    boundaries.sort(_Boundary.compare);
+
+    final _TrieNode trieRoot = _TrieNode();
+    final Set<_Font> currentElements = <_Font>{};
+
+    void newRange(int start, int end) {
+      // Ensure we are using the canonical font order.
+      final List<_Font> fonts = List<_Font>.of(currentElements)
```

Comment:
maybe the algorithm should just always keep currentElements sorted when it inserts a new font into the set

Reply:
I did have it that way but it was too 'clever'. Sorting the boundaries by increasing index for starts but decreasing index for ends, and sorting ends before starts, will give you two ordered subsequences that still need to be merged. When I decided to simply sort, it git rid of ~30 lines of code.